### PR TITLE
feat:  retrieve users' unread notifications

### PR DESCRIPTION
**Pull request by zee-ham-su**
## What does the PR do?
This PR addresses issue of users to retrieve only their unread notifications for authenticated users from the database.  The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done

## How should this be manually tested?
1. Login to a user account, and get the access token

2. You can send a GET request to the following endpoint:
GET /api/v1/notifications.  

## Query Parameters

- is_read=false:

> ```json
> Authorization: Bearer <access_token>
> ```

### Success Response
on a successful execution, you will receive the following response structure:
```
{
  "status": "success",
  "message": "Unread notifications retrieved successfully",
  "status_code": 200,
  "data": {
    "total_notification_count": 5,
    "total_unread_notification_count": 3,
    "notifications": [
      {
        "notification_id": "uuid1",
        "is_read": false,
        "message": "You have a new message",
        "created_at": "2024-07-25T12:00:00Z"
      },
      {
        "notification_id": "uuid2",
        "is_read": false,
        "message": "Your profile has been updated",
        "created_at": "2024-07-24T15:30:00Z"
      }
    ]
  }
}
```

### "Invalid Input" Response
If the request is invalid, you will receive the following response:
```
{
  "status": "error",
  "message": "Invalid request.",
  "status_code": 400
}

```
### Unauthorized Response
```
{
  "status": "error",
  "message": "Unauthorized request",
  "status_code": 401
}
```

### "Internal Server Error" Response

```
{
  "status": "error",
  "message": "Server error",
  "status_code": 500
}

```
###  Manual Testing Instructions

1. Login User:

   - Make a POST request to /api/v1/auth/login with a valid user's email and password.

   - Save the access token from the response.

2. Retrieve Unread Notifications:

    - Make a GET request to /api/v1/notifications?is_read=false with the following:

          Header: Authorization: Bearer <access_token>

3. Verify Response:

      - Ensure the response includes the total count of notifications, total count of unread notifications, and               a list of unread notifications with the correct details.

## Task
- [x]  Implemented GET /api/v1/notifications?is_read=false endpoint in notifications.controller.ts
- [x] Added method to fetch notifications in the NotificationsService.
- [x] Tested endpoint functionality to ensure it handles various scenarios correctly
- [x] Added appropriate HTTP status codes and response messages for edge cases. 
- [x] Added unit tests for the NotificationsService.
- [x] Updated API documentation to reflect the new endpoint

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-74/retrieve-users-unread-notifications)

### CODE REVIEW

File: src/modules/notifications/entities/notifications.entity.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+@Entity()
+export class Notification extends AbstractBaseEntity {
+  @Column({ nullable: false })
+  message: string;
+
+  @Column({ nullable: false, default: false })
+  is_Read: boolean;
+
+  @ManyToOne(() => User, user => user.notifications)
```
Comment (CFCIfe) [inline src/modules/notifications/entities/notifications.entity.ts:13]:
```suggestion
@ManyToOne(() => User, user => user.notifications, { nullable: false, onDelete: 'CASCADE' })
```

File: src/modules/notifications/notifications.controller.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import { UserPayload } from '../user/interfaces/user-payload.interface';
+import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
+
+@ApiBearerAuth()
+@ApiTags('Notifications')
+@Controller('notifications')
+export class NotificationsController {
+  constructor(private readonly notificationsService: NotificationsService) {}
+
+  @Get()
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:12]: Add swagger doc for Ok and error responses. Check other controller files for how to do it.

File: src/modules/notifications/notifications.controller.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+          status_code: 200,
+          data: notifications,
+        };
+      } else {
+        return {
+          status: 'error',
+          message: 'Invalid request',
+          status_code: 400,
+        };
+      }
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:30]:
```suggestion
// Be carefult to confirm the boolean does not return a different value.
if (isRead === "true") {
   // write your else statement
}

// Continue with code in if block
```

File: src/modules/notifications/notifications.controller.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+          message: 'Unread notifications retrieved successfully',
+          status_code: 200,
+          data: notifications,
+        };
+      } else {
+        return {
+          status: 'error',
+          message: 'Invalid request',
+          status_code: 400,
+        };
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:29]:
```suggestion
throw new BadRequestException({
    // wrap all the body in here
})
```

File: src/modules/notifications/notifications.controller.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          message: 'User not found',
+          status_code: 400,
+        };
+      }
+      return {
+        status: 'error',
+        message: 'Failed to retrieve unread notifications',
+        status_code: 500,
+      };
+    }
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:44]: there should be an else statement else the behaviour will be different

File: src/modules/notifications/notifications.service.ts
Comment (CFCIfe) [inline src/modules/notifications/notifications.service.ts:1]: No linting on these files.

run `npm ci`

File: src/modules/notifications/notifications.service.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    private userRepository: Repository<User>
+  ) {}
+
+  async getUnreadNotificationsForUser(userId: string) {
+    try {
+      const user = await this.userRepository.findOne({ where: { id: userId } });
+      if (!user) {
+        throw new NotFoundException('User not found');
+      }
+
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.service.ts:22]: All these logic already exists in the userService.

```suggestion
const user: Partial<UserInterface> = await this.userService.getUserRecord({
      identifier: user_id,
      identifierType: 'id',
    });
```

if no user it is a Bad request not 404... Also test your error responses conform with the issue description, because I can guess this will not

**Review by CFCIfe — COMMENTED** [review]
Review comments, effect changes and Resolve conversations for new review

**Review by davidolanrewaju — APPROVED** [review]

File: src/modules/notifications/notifications.service.ts
Comment (zee-ham-su) [reply src/modules/notifications/notifications.service.ts:1]: resolved

File: .vscode/settings.json
[the remark this answers is not in the document]
Comment (zee-ham-su) [reply .vscode/settings.json:1]: okay

File: src/modules/notifications/notifications.service.ts (hunk 6 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.service.ts:22]: resolved

File: .vscode/settings.json
[the remark this answers is not in the document]
Comment (zee-ham-su) [reply .vscode/settings.json:1]: okay

Comment (CFCIfe) [inline .vscode/settings.json:1]: How to approach this situation is to remove the changes you added to the file.

```,
  "cSpell.words": ["mockuser", "nestjs", "organisation", "organisations", "postbuild"]
```

File: package.json [hunk 7]
```diff
@@ -6,9 +6,10 @@
   "private": true,
   "license": "UNLICENSED",
   "scripts": {
-    "build": "nest build",
+    "build": "node node_modules/@nestjs/cli/bin/nest.js build",
```
Comment (CFCIfe) [inline package.json:9]: No! No! No!.

Revert to

```
nest build
```

File: package.json [hunk 8]
```diff
@@ -6,9 +6,10 @@
...
     "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
     "start": "nest start",
+    "heroku-postbuild": "npm install && npm run build",
```
Comment (CFCIfe) [inline package.json:12]: No!

Take off

File: src/modules/notifications/dtos/unread-responseError.dto.ts
Comment (CFCIfe) [inline src/modules/notifications/dtos/unread-responseError.dto.ts:1]: File name should be `unread-response-error.dto.ts`

File: src/modules/notifications/notifications.controller.ts [hunk 9]
```diff
@@ -39,16 +41,33 @@ export class NotificationsController {
...
-  @ApiOkResponse({ type: MarkAllNotificationAsReadResponse, description: 'Notifications cleared successfully.' })
-  @ApiUnauthorizedResponse({ type: MarkAllNotificationAsReadError, description: 'Unauthorized' })
-  @ApiInternalServerErrorResponse({ type: MarkAllNotificationAsReadError, description: 'Internal Server Error' })
-  @ApiOperation({ summary: 'Marks all notifications a read' })
-  async markAllNotificationsAsRead(@Req() request: Request) {
-    const user = request['user'];
-
-    const userId = user.id;
 
-    return this.notificationsService.markAllNotificationsAsReadForUser(userId);
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:52]: This is someone else's code you have taken off. Please add it back.

```suggestion
 @Delete('/clear')
  @ApiOkResponse({ type: MarkAllNotificationAsReadResponse, description: 'Notifications cleared successfully.' })
  @ApiUnauthorizedResponse({ type: MarkAllNotificationAsReadError, description: 'Unauthorized' })
  @ApiInternalServerErrorResponse({ type: MarkAllNotificationAsReadError, description: 'Internal Server Error' })
  @ApiOperation({ summary: 'Marks all notifications a read' })
  async markAllNotificationsAsRead(@Req() request: Request) {
    const user = request['user'];

    const userId = user.id;

    return this.notificationsService.markAllNotificationsAsReadForUser(userId);
```

File: src/modules/notifications/notifications.controller.ts [hunk 10]
```diff
@@ -39,16 +41,33 @@ export class NotificationsController {
...
+    const userId = req.user.id;
+    if (is_read === 'false') {
+      const notifications = await this.notificationsService.getUnreadNotificationsForUser(userId);
+      return {
+        status: 'success',
+        message: 'Unread notifications retrieved successfully',
+        status_code: 200,
+        data: notifications,
+      };
+    }
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:71]: This method cannot work.

1. is_read is a boolean and should be checked
2. The method should only run when is_read is "false", what happens if is_read is true?

File: src/modules/notifications/notifications.service.ts (hunk 6 above)
Comment (CFCIfe) [reply src/modules/notifications/notifications.service.ts:30]: You haven't resolved. You're still querying the user dB when you don't need to

File: src/modules/notifications/notifications.controller.ts (hunk 10 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:71]: If is_read is true, I can return a response indicating there are no unread notifications to retrieve.

File: src/modules/notifications/notifications.controller.ts (hunk 9 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:52]: Okay

File: .vscode/settings.json
Comment (zee-ham-su) [reply .vscode/settings.json:1]: Okay

Comment (zee-ham-su) [reply .vscode/settings.json:1]: reverted

File: package.json (hunk 7 above)
Comment (zee-ham-su) [reply package.json:9]: suggestion applied

File: package.json (hunk 8 above)
Comment (zee-ham-su) [reply package.json:12]: taken off

File: src/modules/notifications/dtos/unread-responseError.dto.ts
Comment (zee-ham-su) [reply src/modules/notifications/dtos/unread-responseError.dto.ts:1]: applied

File: src/modules/notifications/notifications.service.ts
Comment (CFCIfe) [inline src/modules/notifications/notifications.service.ts:1]: Beautiful changes done. WELL DONE!!!

**Review by CFCIfe — COMMENTED** [review]
LGTM. @alexindevs @Cobidex @rugue and @davidolanrewaju Please review this!

**Review by rugue — DISMISSED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by davidolanrewaju — DISMISSED** [review]

**Review by rugue — APPROVED** [review]

**Review by rugue — APPROVED** [review]

**Review by Cobidex — CHANGES_REQUESTED** [review]

File: src/database/data-source.ts [hunk 11]
```diff
@@ -1,28 +1,28 @@
...
-});
-export async function initializeDataSource() {
-  if (!dataSource.isInitialized) {
-    await dataSource.initialize();
-  }
-  return dataSource;
-}
-
-export default dataSource;
+import { DataSource, DataSourceOptions } from 'typeorm';
```
[the remark this answers is not in the document]
Comment (zee-ham-su) [reply src/database/data-source.ts:1]: You means the data source file ?

**Review by Homoakin619 — APPROVED** [review]

**Review by Homoakin619 — APPROVED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by TheCodeGhinux — DISMISSED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by rugue — APPROVED** [review]

**Review by Cobidex — APPROVED** [review]
